src/test/java/com/flagsmith/FlagsmithCachedClientTest.java

```diff
@@ -0,0 +1,356 @@
+
+    // Act: cache remains with 2 entries
+    environment.client.getUserFlagsAndTraits(user3);
+    clientCache.cleanUp();
+    assertEquals(2, clientCache.estimatedSize());
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_Project_Flags_Then_Not_Cached() {
+    assertEquals(0, clientCache.estimatedSize());
+
+    final List<Flag> featureFlags = environment.client.getFeatureFlags();
+
+    assertEquals(0, clientCache.estimatedSize());
+    assertNotNull(featureFlags);
+    assertThat(featureFlags)
+        .hasSize(2)
+        .containsExactlyInAnyOrder(
+            flag("Flag to be enabled for the user", null, false),
+            flag("Other Flag", null, false)
+        );
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Flags_Then_Reuses_Cache_Entry() {
+    final List<Flag> featureFlags1 = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(featureFlags1);
+
+    final List<Flag> featureFlags1Again = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(featureFlags1Again);
+    assertEquals(featureFlags1, featureFlags1Again);
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Flags_Then_Refetches_Only_If_Not_Present() {
+    final List<Flag> featureFlags1 = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(featureFlags1);
+
+    createFeature(new FlagsmithTestHelper.FlagFeature(
+        "a-new-flag",
+        null,
+        environment.projectId,
+        false));
+
+    final List<Flag> featureFlags1Again = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(featureFlags1Again);
+    assertEquals(featureFlags1, featureFlags1Again);
+
+    clientCache.invalidate(user1.getIdentifier());
+
+    final List<Flag> updatedFlags1 = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(updatedFlags1);
+    assertNotEquals(featureFlags1, updatedFlags1);
+    assertThat(updatedFlags1)
+        .hasSize(3)
+        .containsExactlyInAnyOrder(
+            flag("a-new-flag", null, false),
+            flag("Flag to be enabled for the user", null, false),
+            flag("Other Flag", null, false)
+        );
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Traits_And_Flags_Then_Reuses_Cache_Entry() {
+    final FlagsAndTraits flagsAndTraits1 = environment.client.getUserFlagsAndTraits(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(flagsAndTraits1);
+
+    final FlagsAndTraits flagsAndTraits1Again = environment.client.getUserFlagsAndTraits(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(flagsAndTraits1Again);
+    assertEquals(flagsAndTraits1, flagsAndTraits1Again);
```

Comment:
Added a remote feature flag modification that should not be in the cache.

Reply:
Thanks, that makes sense. So I guess the line `assertEquals(1, clientCache.estimatedSize());` is now checking to confirm that we're using the cache, right? I think it would be good to add a comment / some comments to explain that a little bit.